Source/CsDebugScript.Engine/Engine/Debuggers/DbgEngDll.cs

```diff
@@ -1243,6 +1260,312 @@ public DebugEventInfo GetLastEventInfo()
+                    uint executionStatus = ((IDebugControl7)loopClient).GetExecutionStatus();
+
+                    while (executionStatus == (uint)Defines.DebugStatusBreak)
+                    {
+                        debugStatusBreak.Set();
+                        debugStatusGo.WaitOne();
+
+                        executionStatus = ((IDebugControl7)loopClient).GetExecutionStatus();
+                    }
+
+                    hasClientExited = executionStatus == (uint)Defines.DebugStatusNoDebuggee;
+                }
+            }
+
+            /// <summary>
+            /// Waits for debugger loop thread to finish.
+            /// </summary>
+            public void WaitForDebuggerLoopToExit()
+            {
+                debuggerStateLoop.Join();
+            }
+        }
+
+        /// <summary>
+        /// Debug callbacks called during WaitForEvent callback.
+        /// This class in future can be extended to support callbacks provided
+        /// on certain actions (e.g. breakpoint hit, thread create, module load etc.)
+        /// </summary>
+        class DebugCallbacks : IDebugEventCallbacks
+        {
+            /// <summary>
+            /// IDebugClient.
+            /// </summary>
+            [ThreadStatic]
+            private IDebugClient client;
+
+            /// <summary>
+            /// Constructor.
+            /// </summary>
+            /// <param name="client">IDebugClient interface.</param>
+            /// <param name="debugStatusGoEvent">Event used to signal when debuggee switches to release state.</param>
+            public DebugCallbacks(IDebugClient client, System.Threading.AutoResetEvent debugStatusGoEvent)
+            {
+                this.client = client.CreateClient();
+                this.debugStatusGoEvent = debugStatusGoEvent;
+                this.client.SetEventCallbacks(this);
+            }
+
+            /// <summary>
+            /// Event used to signal go state. Used to unblock DebugFlow loop.
+            /// </summary>
+            private System.Threading.AutoResetEvent debugStatusGoEvent;
+
+            public void Breakpoint(IDebugBreakpoint Bp)
+            {
+                throw new NotImplementedException();
+            }
+
+            /// <summary>
+            /// Callback on change debugee state.
+            /// </summary>
+            /// <param name="Flags"></param>
+            /// <param name="Argument"></param>
+            /// <returns></returns>
+            public void ChangeDebuggeeState(uint Flags, ulong Argument)
+            {
+                uint executionStatus = ((IDebugControl7)client).GetExecutionStatus();
+
+                if (executionStatus == (uint)Defines.DebugStatusGo)
+                {
+                    debugStatusGoEvent.Set();
+                }
+
+                return;
+            }
+
+            public void ChangeEngineState(uint Flags, ulong Argument)
+            {
+                return;
```

Comment:
Removed the return, now is just an empty body. Idea is that in future we may want to provide a way for user to hook up on these events.

---
In reply to: [69631989](https://github.com/southpolenator/WinDbgCs/pull/8#discussion_r69631989) [](ancestors = 69631989)